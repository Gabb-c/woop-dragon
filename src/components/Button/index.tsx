import styles from './styles.module.scss';
interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  label?: string;
  loading?: boolean;
  flavor?: 'delete' | 'save' | 'default';
}

export const Button: React.FC<ButtonProps> = ({
  label,
  flavor,
  loading,
  ...props
}): JSX.Element => {
  return (
    <button
      {...props}
      className={!flavor || flavor === 'default' ? styles.button : styles[`button-${flavor}`]}
    >
      {!loading ? label : 'loading...'}
    </button>
  );
};
