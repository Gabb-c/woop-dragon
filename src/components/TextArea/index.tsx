import styles from './styles.module.scss';

interface TextAreaProps extends React.TextareaHTMLAttributes<HTMLTextAreaElement> {
  label?: string;
  error?: boolean;
  errorText?: string;
}

export const TextArea: React.FC<TextAreaProps> = ({
  label,
  error,
  errorText,
  ...props
}): JSX.Element => {
  return (
    <div className={styles.container}>
      <span className={error ? styles.inputErrorLabel : styles.inputLabel}>{label}</span>
      <textarea {...props} data-testid="textarea-test" className={styles.input} />
      {error && <span className={styles.inputErrorText}>{errorText}</span>}
    </div>
  );
};
