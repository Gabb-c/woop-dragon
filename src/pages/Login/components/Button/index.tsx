import styles from './styles.module.scss';
interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  label?: string;
  loading?: boolean;
  flavor?: 'delete' | 'edit' | 'default';
}

export const Button: React.FC<ButtonProps> = ({ label, loading, ...props }): JSX.Element => {
  return (
    <button {...props} className={styles.button}>
      {!loading ? label : 'loading...'}
    </button>
  );
};
