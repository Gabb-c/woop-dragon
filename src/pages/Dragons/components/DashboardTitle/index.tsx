import styles from './styles.module.scss';
import { Button } from '@components/Button';
import { useAuth } from '@contexts/AuthContext';
import { useNavigate } from 'react-router-dom';

export const DashboardHeader = () => {
  const { signOut } = useAuth();

  const navigate = useNavigate();

  return (
    <div className={styles.container}>
      <Button label="Logout" flavor="delete" onClick={() => signOut()} />
      <h1>Dragons Dashboard</h1>
      <Button label="Add" flavor="save" onClick={() => navigate('/dragons/form')} />
    </div>
  );
};
