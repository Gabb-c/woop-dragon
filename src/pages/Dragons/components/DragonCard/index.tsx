import { Dragon } from '@models/dragon';
import { format } from 'date-fns';
import { Button } from 'src/pages/Login/components/Button';
import styles from './styles.module.scss';

interface DragonCardProps {
  dragon: Dragon;
  onDelete: (id: string) => void;
  onEdit?: (id: string) => void;
}

export const DragonCard: React.FC<DragonCardProps> = ({ dragon, onDelete }): JSX.Element => {
  return (
    <div className={styles.card}>
      <h5 className={styles.dragonType}>{dragon.type.toUpperCase()}</h5>
      <h1 className={styles.dragonName}>{dragon.name}</h1>
      <h3 className={styles.dragonBirth}>{format(new Date(dragon.createdAt), 'MM/dd/yyyy')}</h3>
      <div className={styles.fields}>
        <Button label="Edit" />
        <Button label="Delete" onClick={() => onDelete(dragon.id)} />
      </div>
    </div>
  );
};
