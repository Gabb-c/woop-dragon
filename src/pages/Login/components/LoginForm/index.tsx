import { useAuth } from '@contexts/AuthContext';
import { useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { Button } from '../Button';
import { TextField } from '../TextField';
import styles from './styles.module.scss';

export const LoginForm: React.FC = (): JSX.Element => {
  const { signIn } = useAuth();
  const navigate = useNavigate();

  const [username, setUsername] = useState('');
  const [password, setPassword] = useState('');
  const [error, setError] = useState(false);
  const [loading, setLoading] = useState(false);

  const handleLogin = () => {
    setLoading(true);
    signIn(username, password)
      .then(() => navigate('/dragons'))
      .catch((error_) => {
        setError(true);
        setLoading(false);
        console.error(error_);
      });
  };

  return (
    <div className={styles.card}>
      <h2>Welcome back!</h2>
      <div className={styles.fields}>
        <TextField
          label="Username"
          error={error}
          errorText="Invalid username"
          value={username}
          onChange={(e) => {
            if (error) setError(false);
            setUsername(e.target.value);
          }}
        />
        <TextField
          label="Password"
          error={error}
          errorText="Invalid password"
          value={password}
          type="password"
          onChange={(e) => {
            if (error) setError(false);
            setPassword(e.target.value);
          }}
        />
        <Button label="Login" loading={loading} disabled={loading} onClick={() => handleLogin()} />
      </div>
    </div>
  );
};
