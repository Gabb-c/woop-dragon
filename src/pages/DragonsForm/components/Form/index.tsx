import DragonService from '@services/dragon';
import styles from './styles.module.scss';
import { Button } from '@components/Button';
import { Dragon } from '@models/dragon';
import { PropagateLoader } from 'react-spinners';
import { TextArea } from '@components/TextArea';
import { TextField } from '@components/TextField';
import { useEffect, useState } from 'react';
import { useNavigate, useParams } from 'react-router-dom';

export const Form: React.FC = (): JSX.Element => {
  const navigate = useNavigate();
  const { id } = useParams();

  const [dragon, setDragon] = useState<Dragon>({} as Dragon);
  const [loading, setLoading] = useState(false);
  const [loadingButton, setLoadingButton] = useState(false);

  const handleLoadDragon = async () => {
    if (!id) return;
    setLoading(true);
    await DragonService.getDragonById(id)
      .then((response) => {
        setDragon(response);
        return setLoading(false);
      })
      .catch(() => navigate('/dragons'));
  };

  const handleSubmit = () => {
    setLoadingButton(true);
    if (!id) {
      DragonService.createDragon(dragon)
        .then(() => navigate('/dragons'))
        .catch((_error) => console.error(_error));
    } else {
      DragonService.editDragon(dragon, dragon.id)
        .then(() => navigate('/dragons'))
        .catch((_error) => console.error(_error));
    }
  };

  useEffect(() => {
    handleLoadDragon().catch((_error) => console.error(_error));
  }, []);

  const { name, histories, type } = dragon;

  return (
    <div className={styles.card}>
      <h2>{id ? `Edit Dragon #${id}` : 'Add new Dragon'}</h2>
      <div className={styles.fields}>
        {loading ? (
          <PropagateLoader color="white" />
        ) : (
          <>
            <TextField
              label="Name"
              name="name"
              value={name}
              onChange={({ target: { value } }) => {
                setDragon({ ...dragon, name: value });
              }}
            />
            <TextField
              label="Type"
              name="type"
              value={type}
              onChange={({ target: { value } }) => {
                setDragon({ ...dragon, type: value });
              }}
            />
            <TextArea
              label="Histories"
              name="histories"
              value={histories}
              maxLength={200}
              rows={3}
              onChange={({ target: { value } }) => {
                setDragon({ ...dragon, histories: value });
              }}
            />
            <Button
              label="Save"
              name="save-dragon"
              flavor="save"
              loading={loadingButton}
              disabled={loading}
              onClick={handleSubmit}
            />
            <Button
              label="Cancel"
              name="cancel"
              type="submit"
              flavor="delete"
              onClick={() => navigate('/dragons')}
            />
          </>
        )}
      </div>
    </div>
  );
};
