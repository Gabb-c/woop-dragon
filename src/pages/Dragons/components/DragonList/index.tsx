import DragonService from '@services/dragon';
import styles from './styles.module.scss';
import { DashboardHeader } from '../DashboardTitle';
import { Dragon } from '@models/dragon';
import { DragonCard } from '../DragonCard';
import { useEffect, useMemo, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { TextField } from '@components/TextField';

export const DragonList: React.FC = (): JSX.Element => {
  const [list, setList] = useState<Dragon[]>([]);
  const [loading, setLoading] = useState(true);
  const [filter, setFilter] = useState('');

  const navigate = useNavigate();

  const filtered = useMemo(
    () => list.filter((d) => d.name.toLowerCase().includes(filter.toLowerCase())),
    [filter]
  );

  const loadDragons = async () => {
    setLoading(true);
    await DragonService.getAllDragons()
      .then((response) => {
        const sorted = [...response].sort((a, b) =>
          a.name.localeCompare(b.name, 'en', { sensitivity: 'base' })
        );
        setList(sorted);
        return setLoading(false);
      })
      .catch((error) => console.error(error));
  };

  const handleDelete = (id: string) => {
    DragonService.deleteDragonById(id)
      .then(() => {
        return loadDragons();
      })
      .catch((error) => console.error(error));
  };

  const handleEdit = (id: string) => {
    setLoading(true);
    navigate(`/dragons/form/${id}`);
  };

  useEffect(() => {
    loadDragons().catch((error) => console.error(error));
  }, []);

  return (
    <div className={styles.container}>
      <DashboardHeader />
      <TextField
        placeholder="Search..."
        value={filter}
        onChange={({ target: { value } }) => setFilter(value)}
      />
      {loading ? (
        <>
          <svg width="200px" height="200px" viewBox="0 0 100 100" preserveAspectRatio="xMidYMid">
            <rect x="17.5" y="30" width="15" height="40" fill="#646cff">
              <animate
                attributeName="y"
                repeatCount="indefinite"
                dur="1s"
                calcMode="spline"
                keyTimes="0;0.5;1"
                values="18;30;30"
                keySplines="0 0.5 0.5 1;0 0.5 0.5 1"
                begin="-0.2s"
              ></animate>
              <animate
                attributeName="height"
                repeatCount="indefinite"
                dur="1s"
                calcMode="spline"
                keyTimes="0;0.5;1"
                values="64;40;40"
                keySplines="0 0.5 0.5 1;0 0.5 0.5 1"
                begin="-0.2s"
              ></animate>
            </rect>
            <rect x="42.5" y="30" width="15" height="40" fill="#646cff">
              <animate
                attributeName="y"
                repeatCount="indefinite"
                dur="1s"
                calcMode="spline"
                keyTimes="0;0.5;1"
                values="20.999999999999996;30;30"
                keySplines="0 0.5 0.5 1;0 0.5 0.5 1"
                begin="-0.1s"
              ></animate>
              <animate
                attributeName="height"
                repeatCount="indefinite"
                dur="1s"
                calcMode="spline"
                keyTimes="0;0.5;1"
                values="58.00000000000001;40;40"
                keySplines="0 0.5 0.5 1;0 0.5 0.5 1"
                begin="-0.1s"
              ></animate>
            </rect>
            <rect x="67.5" y="30" width="15" height="40" fill="#646cff">
              <animate
                attributeName="y"
                repeatCount="indefinite"
                dur="1s"
                calcMode="spline"
                keyTimes="0;0.5;1"
                values="20.999999999999996;30;30"
                keySplines="0 0.5 0.5 1;0 0.5 0.5 1"
              ></animate>
              <animate
                attributeName="height"
                repeatCount="indefinite"
                dur="1s"
                calcMode="spline"
                keyTimes="0;0.5;1"
                values="58.00000000000001;40;40"
                keySplines="0 0.5 0.5 1;0 0.5 0.5 1"
              ></animate>
            </rect>
          </svg>
        </>
      ) : (
        <div className={styles.lisCtontainer}>
          {filtered.length > 0
            ? filtered.map((d, i) => (
                <DragonCard
                  onDelete={handleDelete}
                  onEdit={() => handleEdit(d.id)}
                  dragon={d}
                  key={i}
                />
              ))
            : list.map((d, i) => (
                <DragonCard
                  onDelete={handleDelete}
                  onEdit={() => handleEdit(d.id)}
                  dragon={d}
                  key={i}
                />
              ))}
        </div>
      )}
    </div>
  );
};
