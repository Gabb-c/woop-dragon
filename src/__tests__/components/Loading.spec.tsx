import React from 'react';
import { render, screen } from '@testing-library/react';
import { Loading } from '../../components/Loading';
import '@testing-library/jest-dom';

describe('Loading', () => {
  it('test if it renders in the DOM', () => {
    render(<Loading />);

    const resTest = screen.getByTestId(/loading/i);

    expect(resTest).toBeInTheDocument();
  });

  it('should match snapshot', () => {
    render(<Loading />);

    const resTest = screen.getByTestId(/loading/i);

    expect(resTest).toMatchSnapshot();
  });
});
