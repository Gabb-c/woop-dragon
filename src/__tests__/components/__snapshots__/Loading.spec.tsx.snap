// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loading should match snapshot 1`] = `
<svg
  data-testid="loading"
  height="200px"
  preserveAspectRatio="xMidYMid"
  viewBox="0 0 100 100"
  width="200px"
>
  <rect
    fill="#646cff"
    height="40"
    width="15"
    x="17.5"
    y="30"
  >
    <animate
      attributeName="y"
      begin="-0.2s"
      calcMode="spline"
      dur="1s"
      keySplines="0 0.5 0.5 1;0 0.5 0.5 1"
      keyTimes="0;0.5;1"
      repeatCount="indefinite"
      values="18;30;30"
    />
    <animate
      attributeName="height"
      begin="-0.2s"
      calcMode="spline"
      dur="1s"
      keySplines="0 0.5 0.5 1;0 0.5 0.5 1"
      keyTimes="0;0.5;1"
      repeatCount="indefinite"
      values="64;40;40"
    />
  </rect>
  <rect
    fill="#646cff"
    height="40"
    width="15"
    x="42.5"
    y="30"
  >
    <animate
      attributeName="y"
      begin="-0.1s"
      calcMode="spline"
      dur="1s"
      keySplines="0 0.5 0.5 1;0 0.5 0.5 1"
      keyTimes="0;0.5;1"
      repeatCount="indefinite"
      values="20.999999999999996;30;30"
    />
    <animate
      attributeName="height"
      begin="-0.1s"
      calcMode="spline"
      dur="1s"
      keySplines="0 0.5 0.5 1;0 0.5 0.5 1"
      keyTimes="0;0.5;1"
      repeatCount="indefinite"
      values="58.00000000000001;40;40"
    />
  </rect>
  <rect
    fill="#646cff"
    height="40"
    width="15"
    x="67.5"
    y="30"
  >
    <animate
      attributeName="y"
      calcMode="spline"
      dur="1s"
      keySplines="0 0.5 0.5 1;0 0.5 0.5 1"
      keyTimes="0;0.5;1"
      repeatCount="indefinite"
      values="20.999999999999996;30;30"
    />
    <animate
      attributeName="height"
      calcMode="spline"
      dur="1s"
      keySplines="0 0.5 0.5 1;0 0.5 0.5 1"
      keyTimes="0;0.5;1"
      repeatCount="indefinite"
      values="58.00000000000001;40;40"
    />
  </rect>
</svg>
`;
